cmd/policy.go

```diff
@@ -70,42 +89,61 @@ func (sys *PolicySys) IsAllowed(args policy.Args) bool {
 	return args.IsOwner
 }
 
-// Init - initializes policy system from policy.json of all buckets.
-func (sys *PolicySys) Init(objAPI ObjectLayer) error {
-	if objAPI == nil {
-		return errInvalidArgument
-	}
-
+// Refresh PolicySys.
+func (sys *PolicySys) refresh(objAPI ObjectLayer) {
 	buckets, err := objAPI.ListBuckets(context.Background())
 	if err != nil {
-		return err
+		return
 	}
-
+	sys.RemoveDeletedBuckets(buckets)
 	for _, bucket := range buckets {
 		config, err := GetPolicyConfig(objAPI, bucket.Name)
 		if err != nil {
-			if _, ok := err.(BucketPolicyNotFound); !ok {
-				return err
+			if _, ok := err.(ObjectNotFound); ok {
+				// To delete in memory cache of the policy.
+				sys.Set(bucket.Name, policy.Policy{Version: policy.DefaultVersion})
 			}
-		} else {
-			// This part is specifically written to handle migration
-			// when the Version string is empty, this was allowed
-			// in all previous minio releases but we need to migrate
-			// those policies by properly setting the Version string
-			// from now on.
-			if config.Version == "" {
-				logger.Info("Found in-consistent bucket policies, Migrating them for Bucket: (%s)", bucket.Name)
-				config.Version = policy.DefaultVersion
-
-				if err = savePolicyConfig(objAPI, bucket.Name, config); err != nil {
-					return err
-				}
+			continue
+		}
+		// This part is specifically written to handle migration
+		// when the Version string is empty, this was allowed
+		// in all previous minio releases but we need to migrate
+		// those policies by properly setting the Version string
+		// from now on.
+		if config.Version == "" {
```

Comment:
This check should be applicable only in `Init()`.  Its inefficient to do on every refresh.

Reply:
Since it is just a string comparison, it is not expensive. But it will help readability slightly because why will we want to migrate during refresh when we would have already migrated during init. But fixing it will cause a bit of duplication of code and hence i have left it this way.